Mathlib/NumberTheory/ClassNumber/Finite.lean

```diff
@@ -0,0 +1,429 @@
+theorem exists_min (I : (Ideal S)⁰) :
+    ∃ b ∈ (I : Ideal S),
+      b ≠ 0 ∧ ∀ c ∈ (I : Ideal S), abv (Algebra.norm R c) < abv (Algebra.norm R b) → c =
+      (0 : S) := by
+  classical
+  obtain ⟨_, ⟨b, b_mem, b_ne_zero, rfl⟩, min⟩ := @Int.exists_least_of_bdd
+      (fun a => ∃ b ∈ (I : Ideal S), b ≠ (0 : S) ∧ abv (Algebra.norm R b) = a) _
+    (by
+      use 0
+      rintro _ ⟨b, _, _, rfl⟩
+      apply abv.nonneg)
+    (by
+      obtain ⟨b, b_mem, b_ne_zero⟩ := (I : Ideal S).ne_bot_iff.mp (nonZeroDivisors.coe_ne_zero I)
+      exact ⟨_, ⟨b, b_mem, b_ne_zero, rfl⟩⟩)
+  · refine' ⟨b, b_mem, b_ne_zero, _⟩
+    intro c hc lt
+    contrapose! lt with c_ne_zero
+    exact min _ ⟨c, hc, c_ne_zero, rfl⟩
+#align class_group.exists_min ClassGroup.exists_min
+
+section IsAdmissible
+
+variable {abv} (adm : abv.IsAdmissible)
+
+/-- If we have a large enough set of elements in `R^ι`, then there will be a pair
+whose remainders are close together. We'll show that all sets of cardinality
+at least `cardM bS adm` elements satisfy this condition.
+
+The value of `cardM` is not at all optimal: for specific choices of `R`,
+the minimum cardinality can be exponentially smaller.
+-/
+noncomputable def cardM : ℕ :=
+  adm.card (normBound abv bS ^ (-1 / Fintype.card ι : ℝ)) ^ Fintype.card ι
+set_option linter.uppercaseLean3 false in
+#align class_group.cardM ClassGroup.cardM
+
+variable [Infinite R]
+
+/-- In the following results, we need a large set of distinct elements of `R`. -/
+noncomputable def distinctElems : Fin (cardM bS adm).succ ↪ R :=
+  Fin.valEmbedding.trans (Infinite.natEmbedding R)
+#align class_group.distinct_elems ClassGroup.distinctElems
+
+variable [DecidableEq R]
+
+/-- `finset_approx` is a finite set such that each fractional ideal in the integral closure
+contains an element close to `finset_approx`. -/
+noncomputable def finsetApprox : Finset R :=
+  (Finset.univ.image fun xy : _ × _ => distinctElems bS adm xy.1 - distinctElems bS adm xy.2).erase
+    0
+#align class_group.finset_approx ClassGroup.finsetApprox
+
+theorem finsetApprox.zero_not_mem : (0 : R) ∉ finsetApprox bS adm :=
+  Finset.not_mem_erase _ _
+#align class_group.finset_approx.zero_not_mem ClassGroup.finsetApprox.zero_not_mem
+
+@[simp]
+theorem mem_finsetApprox {x : R} :
+    x ∈ finsetApprox bS adm ↔ ∃ i j, i ≠ j ∧ distinctElems bS adm i - distinctElems bS adm j =
+    x := by
+  simp only [finsetApprox, Finset.mem_erase, Finset.mem_image]
+  constructor
+  · rintro ⟨hx, ⟨i, j⟩, _, rfl⟩
+    refine' ⟨i, j, _, rfl⟩
+    rintro rfl
+    simp at hx
+  · rintro ⟨i, j, hij, rfl⟩
+    refine' ⟨_, ⟨i, j⟩, Finset.mem_univ _, rfl⟩
+    rw [Ne.def, sub_eq_zero]
+    exact fun h => hij ((distinctElems bS adm).injective h)
+#align class_group.mem_finset_approx ClassGroup.mem_finsetApprox
+
+section Real
+
+open Real
+
+attribute [-instance] Real.decidableEq
+
+set_option maxHeartbeats 800000 in
```

Comment:
This proof is rather slow. Maybe @Vierkantor wants to have a look at it (but it is already slow in mathlib3).